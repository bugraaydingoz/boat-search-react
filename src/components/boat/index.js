export { Boat } from './Boat'
export { BoatList } from './BoatList'
