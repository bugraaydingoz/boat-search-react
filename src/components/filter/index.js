export { Filter } from './Filter'
export { FilterList } from './FilterList'
