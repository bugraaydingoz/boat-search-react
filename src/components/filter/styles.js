import styled from 'styled-components'
import { STYLES } from '../shared/styles'

export const FilterListContainer = styled.div`
  display: flex;
  margin: auto;
  width: 740px;
`

export const FilterContainer = styled.div`
  display: flex;
  flex-direction: column;
  margin-right: 20px;
  padding: 5px 15px 5px 25px;
  width: 120px;
  /* height: 45px; */
  border: 1px solid #e1e1e5;
  border-radius: 999px;
`

export const Label = styled.label`
  margin: 0;
  display: flex;
  font-weight: normal;
  font-size: 12px;
  line-height: 18px;
  color: ${STYLES.color.gray[600]};
`

export const InputContainer = styled.div`
  display: flex;
`

export const Input = styled.input`
  display: flex;
  width: 32px;
  font-weight: 500;
  font-size: 12px;
  line-height: 18px;
  color: ${STYLES.color.primary.default};
  border: none;
`
