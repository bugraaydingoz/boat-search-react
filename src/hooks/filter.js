const itemsPerPage = 10

function getUrlValues(urlQuery) {
  const { q, minLength, maxLength, minYear, maxYear } = urlQuery
  return {
    q: q ? q : null,
    minLength: minLength ? Number(minLength) : null,
    maxLength: maxLength ? Number(maxLength) : null,
    minYear: minYear ? Number(minYear) : null,
    maxYear: maxYear ? Number(maxYear) : null,
  }
}

function filterByText(q, boat) {
  const { name, type, locality, country } = boat
  const textFields = [
    name?.toLowerCase(),
    type?.toLowerCase(),
    locality?.toLowerCase(),
    country?.toLowerCase(),
  ]
  return !textFields.some((t) => t.includes(q.toLowerCase()))
}

export function useFilter(boats = [], urlQuery) {
  const { q, minLength, maxLength, minYear, maxYear } = getUrlValues(urlQuery)

  const filteredBoats = boats.filter((boat) => {
    if (q && filterByText(q, boat)) {
      return false
    }

    if (minLength && boat.length < minLength) {
      return false
    }

    if (maxLength && boat.length > maxLength) {
      return false
    }

    if (minYear && boat.year < minYear) {
      return false
    }

    if (maxYear && boat.year > maxYear) {
      return false
    }

    return true
  })

  const paginatedBoats = filteredBoats.slice(
    Number(urlQuery.page - 1) * itemsPerPage,
    Number(urlQuery.page) * itemsPerPage
  )

  const totalPages = Math.ceil(filteredBoats.length / itemsPerPage)

  return { paginatedBoats, totalPages, length: filteredBoats.length }
}
