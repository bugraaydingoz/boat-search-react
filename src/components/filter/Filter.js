import React, { useState, useEffect } from 'react'

// Styles
import { FilterContainer, Label, Input, InputContainer } from './styles'

export function Filter({ name, value, onChange }) {
  const [min, setMin] = useState('')
  const [max, setMax] = useState('')

  useEffect(() => {
    if (!!value.min) {
      setMin(value.min)
    }
    if (!!value.min) {
      setMax(value.max)
    }
  }, [value])

  return (
    <FilterContainer>
      <Label>{name}</Label>
      <InputContainer>
        <Input
          type="text"
          value={min}
          onChange={(e) => setMin(e.target.value)}
          onBlur={() => onChange({ min, max })}
          onKeyPress={(e) => e.key === 'Enter' && onChange({ min, max })}
        />{' '}
        -{' '}
        <Input
          type="text"
          value={max}
          onChange={(e) => setMax(e.target.value)}
          onBlur={() => onChange({ min, max })}
          onKeyPress={(e) => e.key === 'Enter' && onChange({ min, max })}
        />
      </InputContainer>
    </FilterContainer>
  )
}
