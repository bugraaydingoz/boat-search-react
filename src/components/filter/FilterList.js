import React from 'react'

// Styles
import { FiltersContainer } from './styles'

export function FilterList({ children }) {
  return <FiltersContainer>{children}</FiltersContainer>
}
