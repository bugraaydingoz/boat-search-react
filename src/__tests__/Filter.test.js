import React from 'react'
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { Filter } from '../components/filter'

test('has correct label', () => {
  const label = 'length'
  const { getByTestId } = render(
    <Filter name={label} value={{ min: 10, max: 20 }} />
  )

  expect(getByTestId('filter-label').innerHTML).toEqual(label)
})

test('renders correctly', () => {
  const tree = renderer
    .create(<Filter name="Length" value={{ min: 10, max: 20 }} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
