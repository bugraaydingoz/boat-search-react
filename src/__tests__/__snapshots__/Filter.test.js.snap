// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-AxiKw ioCqkX"
>
  <label
    className="sc-AxhCb fdKtUo"
    data-testid="filter-label"
  >
    Length
  </label>
  <div
    className="sc-AxhUy gWaiz"
  >
    <input
      className="sc-AxgMl DnjmU"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyPress={[Function]}
      type="text"
      value=""
    />
    <span
      className="sc-AxheI fUhWMh"
    >
      -
    </span>
    <input
      className="sc-AxgMl DnjmU"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyPress={[Function]}
      type="text"
      value=""
    />
  </div>
</div>
`;
